policy_maker/app.py

```diff
@@ -0,0 +1,198 @@
+      tag_value = tag.get('Value')
+      return tag_value
+  else:
+    raise ValueError(f'Could not derive a value for {principal_tag_name} from tags')
+
+
+def combine_principals(principal, extra_principals):
+  principals = [principal]
+  if extra_principals is not None:
+    principals = extra_principals + principals
+  return principals
+
+
+def create_policy_document(bucket_name, principals, require_encrypted=False):
+  bucket_arn = f'arn:aws:s3:::{bucket_name}'
+  bucket_objects_arn = f'{bucket_arn}/*'
+  policy = {
+    'Version': '2012-10-17',
+    'Statement': [
+      {
+        'Sid': 'ReadAccess',
+        'Effect': 'Allow',
+        'Principal': {
+          'AWS': principals
+        },
+        'Action': [
+          's3:ListBucket*',
+          's3:GetBucketLocation'
+        ],
+        'Resource': bucket_arn
+      },
+      {
+        'Sid': 'WriteAccess',
+        'Effect': 'Allow',
+        'Principal': {
+          'AWS': principals
+        },
+        'Action': [
+          's3:*Object*',
+          's3:*MultipartUpload*'
+        ],
+        'Resource': bucket_objects_arn
+      }
+    ]
+  }
+  if require_encrypted:
+    additional_statements = [
+      {
+        'Sid': 'DenyIncorrectEncryptionHeader',
+        'Effect': 'Deny',
+        'Principal': {
+          'AWS': principals
+        },
+        'Action': 's3:PutObject',
+        'Resource': bucket_objects_arn,
+        'Condition': {
+          'StringNotEquals': {
+            's3:x-amz-server-side-encryption': 'AES256'
+          }
+        }
+      },
+      {
+        'Sid': 'DenyUnEncryptedObjectUploads',
+        'Effect': 'Deny',
+        'Principal': {
+          'AWS': principals
+        },
+        'Action': 's3:PutObject',
+        'Resource': bucket_objects_arn,
+        'Condition': {
+          'Null': {
+            's3:x-amz-server-side-encryption': 'true'
+          }
+        }
+      }
+    ]
+    policy['Statement'] = policy['Statement'] + additional_statements
+  return json.dumps(policy)
+
+
+def attach_policy(bucket_name, policy_document):
+  '''Add a policy to the bucket'''
+  s3 = get_client('s3')
+  try:
+    s3.put_bucket_policy(Bucket=bucket_name, Policy=policy_document)
+  except ClientError as e:
+    log.debug(e.response['Error']['Message'])
+    raise e
+
+
+@helper.update
+@helper.create
+def create(event, context):
+  '''Create or re-create the policy'''
+  bucket_name, extra_principals, require_encryption = get_params(event)
+  tags = get_bucket_tags(bucket_name)
+  principal_arn = get_principal_tag(tags)
+  principals = combine_principals(
+    principal=principal_arn,
+    extra_principals=extra_principals)
+  policy_document = create_policy_document(
+    bucket_name=bucket_name,
+    principals=principals,
+    require_encrypted=require_encryption)
+  attach_policy(bucket_name, policy_document)
+  aws_request_id = context.aws_request_id
+  physical_resource_id = f'BucketPolicy_{bucket_name}_{aws_request_id}'
```

Comment:
sorry, i missed this last time.  Why are you creating your own PhysicalResourceId?  I believe CFN creates one for you in the response?  This is what a create response looks like for my custom resource:
```
[DEBUG]	2020-04-30T20:34:28.490Z	50113656-356a-4259-9597-eb5770aeae88
{
    "Status": "SUCCESS",
    "PhysicalResourceId": "khai-sc-product-ec2-linux-jumpcloud_AssociateRestartEC2InstanceAction_X1IHCX70",
    "StackId": "arn:aws:cloudformation:us-east-1:465877038949:stack/khai-sc-product-ec2-linux-jumpcloud/9992fc70-8b21-11ea-9b4a-1271b8f35690",
    "RequestId": "c4e77e83-e361-4c6e-850c-8c84e245978b",
    "LogicalResourceId": "AssociateRestartEC2InstanceAction",
    "Reason": "",
    "Data": {}
}
```

Reply:
It's optional to define your own. I don't see a reason not to keep this.

Reply:
sounds good